### ISSUE TITLE
Add HTTP Keep Alive support to Kitura-net

### ISSUE THREAD

**Issue by shmuelk** [IBM-Swift/Kitura#435]

**Comment by djones6** [comment]
Now that we've overcome the memory leak ( #419 ) I've been experimenting with load on Mac, and have some interesting observations.

Driving Kitura at thousands of requests per second more-or-less saturates the CPU on the system.  However, I get periods where throughput drops off, and my console connections begin to lag.

Using 'htop' (available through brew) running via sudo,  I found that a lot of time is being spent in the 'kernel-task' and 'symptomsd' processes during (and persisting for a short while after) driving load against Kitura.  There's precious little information about what 'symptomsd' actually is, but it seems to be networking related.  My guess is that the activity in these processes (and my console issues) is related to the constant opening + closing of connections due to not having keepalive support.

Some more supporting evidence is that if I perform the same load test with Zewo, albeit running only single-threaded but achieving higher reqests/sec, there is no activity at all in 'symptomsd' and 'kernel-task' is much lower.
If I then drive Zewo specifying the header -H "Connection: close" from wrk, I see similar behaviour to Kitura (lower throughput + lots of symptomsd activity).

Therefore I suspect the benefit of keepalive will be higher on Mac than Linux.

**Comment by shmuelk** [comment]
I am trying a completely different approach.

**Comment by djones6** [comment]
I just posted performance numbers to #588, but copying the relevant numbers here:
```
Implementation | Avg Throughput | Max Throughput | Avg CPU | Avg RSS (kb)
---------------|----------------|----------------|---------|--------------
       default |        37159.7 |       37953.08 |    83.2 |       843132
         async |        53968.0 |       54205.75 |    91.7 |       959214
```

In my initial attempts to run the asynchronous branch with the reset() fix, I had problems with the process apparently dying (though it produced no error message nor core file) when running through my measurement script.  I couldn't reproduce that problem running it manually, however.
On one (and so far, only one) occasion I saw the message `Error occurred on a file descriptor of an epool wait` (sic) printed to the console, and it was not fatal.
Today, I've been able to run multiple times against the asynchronous branch with no further issues.  I've rebuilt Kitura since yesterday, so I wonder if there was some other problem the first time I built it.  I will do a few longer runs with it today to get some more confidence.

**Comment by shmuelk** [comment]
In trying to migrate the asynchronous branch to 06/20, I came across the error described here: https://bugs.swift.org/browse/SR-2050

**Comment by djones6** [comment]
Performance numbers for the asynchronous and gcd_async branches:
4-way
```
Implementation | Avg Throughput | Max Throughput | Avg CPU | Avg RSS (kb)
---------------|----------------|----------------|---------|--------------
   Kitura 0.21 |        35843.6 |       36571.65 |    83.1 |       813018
     gcd_async |         9394.1 |        9796.21 |    57.9 |       260384
 gcd_async+fix |        38122.0 |       38609.66 |    88.6 |       793990
         epoll |        50714.5 |       51524.28 |    91.2 |       900478
    vapor 0.12 |        33068.6 |       33508.57 |    99.9 |        53444
 zewo 0.7(oob) |         3196.4 |        3196.90 |    10.1 |        42287
 zewo 0.7(max) |        44218.3 |       44836.83 |    99.7 |       112296
```
8-way
```
Implementation | Avg Throughput | Max Throughput | Avg CPU | Avg RSS (kb)
---------------|----------------|----------------|---------|--------------
   Kitura 0.21 |        53782.2 |       54259.25 |    72.0 |      1211240
     gcd_async |        13727.4 |       14139.34 |    45.4 |       336980
 gcd_async+fix |        66948.0 |       67792.57 |    86.8 |      1314806
         epoll |        90269.3 |       91668.39 |    87.1 |      1584578
    vapor 0.12 |        61156.4 |       61491.64 |    99.9 |        54025
 zewo 0.7(oob) |         3194.0 |        3199.32 |     4.5 |        78319
 zewo 0.7(max) |        82478.3 |       82721.29 |    99.6 |       151542
```
'gcd_async+fix' refers to the following change to `Sources/KituraNet/IncomingHTTPSocketHandler.swift`:
```
-        dispatch_io_read(channel!, 0, Int.max, HTTPServer.clientHandlerQueue.osQueue) {done, data, error in
+        dispatch_io_read(channel!, 0, 1*1024, HTTPServer.clientHandlerQueue.osQueue) {done, data, error in
```

**Comment by shmuelk** [comment]
Merged.
